useradd: Do not automatically add supplements groups for system users

SgAkErRu opened the pull request:
Do not automatically add supplements groups for system users that is created via the `-r` flag. 

Before this change, groups from the `GROUPS` parameter from the `/etc/default/useradd` file were automatically added for the system user, which is most often undesirable (more precisely, I do not even know the case when it would be necessary).

Related to: https://github.com/shadow-maint/shadow/pull/586/

SgAkErRu commented:
> Could you please explain why we want this?
> 
> Isn't this another way of having an empty GROUPS list? Why not have an empty GROUPS list instead? (I'm not sure I liked the GROUPS list in the first place.)

`GROUPS` allows maintainers to set a list of additional (supplementary) default groups, for example `users`, `dialout`, `audio`, `serial` and so on.

Or, for example, only `users`, which is then expanded through the `libnss-role`, i.e. these groups (`dialout`, `audio`, `serial`) are added through `libnss-role` to those users who are in the `users` group.

But at this moment, it works so that when creating a **system** user via `useradd --system` or `useradd -r`, these groups are also added to it, which is most often undesirable (more precisely, I do not even know the case when it would be necessary).

Therefore, this new option was created, which allows maintainers to disable auto-addition of additional groups for **system** users.

alejandro-colomar commented on src/useradd.c:
```diff
@@ -849,6 +846,18 @@ static int get_groups (char *list)
 	return 0;
 }
 
+/*
+ * free_groups - free current user_groups list
+ */
+static void free_groups (void)
+{
+	int i = 0;
+	while (NULL != user_groups[i]) {
+		free(user_groups[i]);
+		user_groups[i++] = NULL;
+	}
+}
+
```
This looks like a generic function for freeing a (NULL-terminated) list.  We have list interfaces in <lib/list.c>.

Would you mind calling this API free_list() and moving it to lib/list.c?

Also, the loop looks more like a for loop IMO.  How about this?

```c
void
free_list(char **ls)
{
	for (size_t i = 0; ls != NULL; i++) {
		free(ls[i]);
		ls[i] = NULL;
	}
}
```

SgAkErRu replied:
I don't mind. 

Should adding the `free_list()` function and using it here in `useradd.c` be in the same PR or in two different ones?

alejandro-colomar replied:
You can do it in the same PR, in a separate commit.

alejandro-colomar replied:
Oh, and it can be simplified into this:

```c
void
free_list(char **ls)
{
	for (size_t i = 0; ls != NULL; i++)
		free(ls[i]);
	ls[0] = NULL;
}

alejandro-colomar commented on src/useradd.c:
```diff
@@ -764,11 +765,7 @@ static int get_groups (char *list)
-	while (NULL != user_groups[i]) {
-		free(user_groups[i]);
-		user_groups[i++] = NULL;
-	}
+	free_groups ();
```
I would prefer passing the list, to reduce the use of global variables.

alejandro-colomar replied:
Please do this refactor in a preparation patch, so that the commit that adds the feature is minimal.

ikerexxe commented:
> But at this moment, it works so that when creating a **system** user via `useradd --system` or `useradd -r`, these groups are also added to it, which is most often undesirable (more precisely, I do not even know the case when it would be necessary).

Since you are the original author of this functionality and you don't know the case, why don't you filter the group membership addition for non-system users instead of creating a new option? I think that would be the most sensible approach.

SgAkErRu commented:
> > But at this moment, it works so that when creating a **system** user via `useradd --system` or `useradd -r`, these groups are also added to it, which is most often undesirable (more precisely, I do not even know the case when it would be necessary).
> 
> Since you are the original author of this functionality and you don't know the case, why don't you filter the group membership addition for non-system users instead of creating a new option? I think that would be the most sensible approach.

Because I am not original author of this functionality. 

So, that's why I suggest this option, which will keep the possibility of the original behavior (when these groups automatically adds to system users).

ikerexxe commented:
> Because I am not original author of this functionality.

Ups sorry, I have mixed up the names.

@AZaugg any opinion on the following topic?

> > But at this moment, it works so that when creating a **system** user via `useradd --system` or `useradd -r`, these groups are also added to it, which is most often undesirable (more precisely, I do not even know the case when it would be necessary).
> 
> Since you are the original author of this functionality and you don't know the case, why don't you filter the group membership addition for non-system users instead of creating a new option? I think that would be the most sensible approach.

SgAkErRu committed:
lib/list.c: free_list(): Add function

SgAkErRu committed:
src/useradd.c: Use free_list() to free user groups list

alejandro-colomar commented on src/useradd.c:
```diff
@@ -764,11 +764,7 @@ static int get_groups (char *list)
-		free(user_groups[i]);
-		user_groups[i++] = NULL;
-	}
+	free_list (user_groups);
```
Please don't add that whitespace.

alejandro-colomar commented on src/useradd.c:
```diff
@@ -1591,6 +1591,13 @@ static void process_flags (int argc, char **argv)
+	} else {
+		/* If SYS_USER_AUTO_GROUPS_ENAB is disabled,
+		 * then do not automatically add supplements groups for system users. */
+		if (!getdef_bool ("SYS_USER_AUTO_GROUPS_ENAB") && !Gflg && do_grp_update) {
```
Please don't add whitespace here in `getdef_bool (` either.  I know the old code has it, but I'm working on slowly getting rid of that.

Thanks!

alejandro-colomar replied:
Thanks!

ikerexxe commented:
@SgAkErRu yes, go ahead

SgAkErRu committed:
src/useradd.c: Do not automatically add supplements groups for system users

SgAkErRu committed:
useradd: fix test `69_useradd_default_GROUPS_name`

Missing files have been added for the test to work: `gshadow`, `passwd`, `shadow`.
Without them, the foo user was created with a different UID and thus the test failed.

And other minor improvements, such as removing extra spaces and adding empty lines.

SgAkErRu committed:
useradd: tests for supplements groups

For regular and system user cases

ikerexxe reviewed (changes requested):
The current implementation contains a flaw. The `grp_add()` function also sets `do_grp_update = true` in [line 1950](https://github.com/shadow-maint/shadow/blob/master/src/useradd.c#L1950). That function is called when `useradd` is called with `-U` flag. That means that if a user is created with `-U` and without `-G` flags, then the logic you implemented will be executed. That's an undesired execution and it will leave some users without their default supplementary groups.

In addition, I added an inline comment. Probably the end code won't look like that once you fix the aforementioned flaw but it should be something similar.

Finally, one of the commits mentions `useradd: fix test 69_useradd_default_GROUPS_name`, what is it exactly you are fixing? I'm assuming you are adapting to the new GROUPS behaviour you are trying to implement, but I'd make it more obvious in the commit message.

ikerexxe reviewed (approved):
LGTM. Thank you for the patches
